src/include/daos/task.h

```diff
@@ -108,8 +108,8 @@ dc_obj_list_recx_task_create(daos_handle_t oh, daos_handle_t th,
 			     tse_sched_t *tse, tse_task_t **task);
 int
 dc_obj_list_obj_task_create(daos_handle_t oh, daos_handle_t th,
-			    daos_key_t *dkey, daos_key_t *akey,
-			    daos_size_t *size, uint32_t *nr,
+			    daos_epoch_t *epoch, daos_key_t *dkey,
```

Comment:
What's this new parameter '*epoch' for?

Reply:
The leader replica will return current highest epoch for the first list, then the subsequent list on other replicas will skip higher epoch than such one. If there is concurrent modification during the consistency verification, because DAOS does not support DLM lock, we have to retry the list/fetch to check whether some 'inconsistency' between two replicas are caused by race modification or not. Such epoch will avoid falling into retry for ever if there are continuously update against the target.

Reply:
I don't quite see your logic here:
- This is object list interface,  the leader could be different for different dkeys.
- Why the epochs on other replicas higher than the 'max epoch' returned by listing on leader can be ignored? That's the inconsistency we try to detect here.

The 'retry for inconsistency caused by concurrent modification' doesn't make sense to me. I think we can just assume this feature will only be used on a quiescent system.

Reply:
> This is object list interface, the leader could be different for different dkeys.

Currently, for a given object, the dkeys in the same redundancy group have the same leader.

> Why the epochs on other replicas higher than the 'max epoch' returned by listing on leader can be ignored? That's the inconsistency we try to detect here.

The leader returns the latest epoch, that is higher than any completed and in-processed medication's epoch. Use such epoch as the list epoch. Then if there is newer modification after the list start, its epoch will be higher than the list one, so can be ignored.

> The 'retry for inconsistency caused by concurrent modification' doesn't make sense to me. I think we can just assume this feature will only be used on a quiescent system.

For a DAOS user, it is not easy to guarantee that. For example, assume you are accessing some file on a distributed filesystem, if you do not explicitly hold some lock, how can you guarantee that nobody will access the same file by race? We should offer some robust tools, rather than define very strict conditions for such tools use scenario.

Reply:
yes I do think we can rely on the epoch returned by the leader, this is the reason you've removed need_retry & has_retried?